### PR TITLE
feat: integrating signature decoding api

### PR DESCRIPTION
**Pull request by jpuri**
## Explanation

Integrate signature decoding api in signature-controller. When a permit request is received this api is called and the result is asynchronously updated into signature request.

## References

* Fixes https://github.com/MetaMask/MetaMask-planning/issues/3553

### `@metamask/signature-controller`

- **Added**: For Permit signatures invoke signature decoding api and update the signature with the response.

## Checklist

- [x] I've updated the test suite for new or updated code as appropriate
- [x] I've updated documentation (JSDoc, Markdown, etc.) for new or updated code as appropriate
- [ ] I've highlighted breaking changes using the "BREAKING" category above as appropriate
- [ ] I've prepared draft pull requests for clients and consumer packages to resolve any breaking changes

### CODE REVIEW

File: packages/signature-controller/src/SignatureController.test.ts [hunk 1]
```diff
@@ -890,6 +911,23 @@ describe('SignatureController', () => {
...
+        PERMIT_REQUEST_MOCK,
+        SignTypedDataVersion.V4,
+        { parseJsonData: false },
+      );
+
+      await flushPromises();
+      expect(controller.state.signatureRequests[ID_MOCK].decodedRequest).toBe(
+        'IN_PROGRESS',
+      );
+    });
```
Comment (jpuri) [inline packages/signature-controller/src/SignatureController.test.ts:930]: I could not get this test to work, any suggestion here will be useful.

File: packages/signature-controller/src/SignatureController.ts [hunk 2]
```diff
@@ -880,4 +897,67 @@ export class SignatureController extends BaseController<
...
+    chainId: string,
+  ) {
+    // Code below will invoke signature request decoding api for permits
+    if (!this.#decodingApi) {
+      return;
+    }
+    let decodedRequest: DecodedRequestInfo;
+    try {
+      const { primaryType } = JSON.parse(request.params?.[1] ?? '');
+      if (primaryType === 'Permit') {
```
Comment (matthewwalsh0) [inline packages/signature-controller/src/SignatureController.ts:913]: Why are we limiting this to a specific primary type? Even if the API didn't support a type, should it not just return an empty response?

We don't want to maintain this here for every primary type it supports.

File: packages/signature-controller/src/SignatureController.ts [hunk 3]
```diff
@@ -880,4 +897,67 @@ export class SignatureController extends BaseController<
...
+        this.#updateMetadata(signatureRequestId, (draftMetadata) => {
+          draftMetadata.decodedRequest = 'IN_PROGRESS';
```
Comment (matthewwalsh0) [inline packages/signature-controller/src/SignatureController.ts:915]: Rather than having a property with an alternating type, should we provide an explicit `decodingLoading` property instead and leave this undefined?

File: packages/signature-controller/src/utils/normalize.ts [hunk 4]
```diff
@@ -59,3 +59,16 @@ function normalizePersonalMessageData(data: string) {
...
   return bytesToHex(Buffer.from(data, 'utf8'));
 }
+
+/**
+ * Takes a stringified JSON and replaces all numeric values in it with quoted strings.
+ *
+ * @param str - String of JSON to be fixed.
+ * @returns String with all numeric values converted to quoted strings.
+ */
+export function convertNumbericValuestoQuotedString(str?: string) {
```
Comment (matthewwalsh0) [inline packages/signature-controller/src/utils/normalize.ts:69]: Why is this required? Why are numbers not supported in the API?

Comment (matthewwalsh0) [inline packages/signature-controller/src/utils/normalize.ts:69]:
```suggestion
export function convertNumericValuesToQuotedString(str?: string) {
```

File: packages/signature-controller/src/SignatureController.ts [hunk 5]
```diff
@@ -880,4 +897,67 @@ export class SignatureController extends BaseController<
...
+                convertNumbericValuestoQuotedString(params?.[1]) ?? '',
+              ),
+            ],
+          }),
+          headers: { 'Content-Type': 'application/json' },
+        })
+          .then((response) => {
+            return response.json();
+          })
+          .then((result) => {
```
Comment (matthewwalsh0) [inline packages/signature-controller/src/SignatureController.ts:935]: Rather than relying on `then` and `catch`, would it be easier for this method to be `async`?

File: packages/signature-controller/src/SignatureController.ts [hunk 6]
```diff
@@ -880,4 +897,67 @@ export class SignatureController extends BaseController<
...
+              draftMetadata.decodedRequest = decodedRequest;
+            });
+          });
+      }
+    } catch (error) {
+      this.#updateMetadata(signatureRequestId, (draftMetadata) => {
+        draftMetadata.decodedRequest = {
+          error: {
+            message: error as string,
+            type: 'DECODING_FAILED_WITH_ERROR',
```
Comment (matthewwalsh0) [inline packages/signature-controller/src/SignatureController.ts:957]: Should we define an enum such as `DecodingErrorType` for these?

File: packages/signature-controller/src/types.ts [hunk 7]
```diff
@@ -15,6 +15,12 @@ export type OriginalRequest = {
 
   /** Response following a security scan of the request. */
   securityAlertResponse?: Record<string, Json>;
+
+  /** Method of signature request */
+  method?: string;
```
Comment (matthewwalsh0) [inline packages/signature-controller/src/types.ts:20]: Minor, alphabetical?

File: packages/signature-controller/src/types.ts [hunk 8]
```diff
@@ -71,11 +77,29 @@ export type MessageParamsTyped = MessageParams & {
         primaryType: string;
         message: Json;
       };
-
+  /** Signature method V1, V3 or V4 */
+  signatureMethod?: string;
```
Comment (matthewwalsh0) [inline packages/signature-controller/src/types.ts:81]: Where is this populated in the message params?

We're trying to move away from adding properties to the params that are more metadata and stored in the underlying `SignatureRequest`.

File: packages/signature-controller/src/types.ts [hunk 9]
```diff
@@ -71,11 +77,29 @@ export type MessageParamsTyped = MessageParams & {
...
   /** Version of the signTypedData request. */
   version?: string;
 };
 
+export type DecodedRequestInfo =
```
Comment (matthewwalsh0) [inline packages/signature-controller/src/types.ts:86]: Should we avoid coupling to the API and call this `DecodingData` as a parallel to `SimulationData` in the transaction controller?

File: packages/signature-controller/src/types.ts [hunk 10]
```diff
@@ -71,11 +77,29 @@ export type MessageParamsTyped = MessageParams & {
...
+  | {
+      assetType: string;
+      changeType: string;
+      address: string;
+      amount: string;
+      contractAddress: string;
+    }
+  | {
```
Comment (matthewwalsh0) [inline packages/signature-controller/src/types.ts:94]: Rather than defining these as a union of alternate types, could we define additional types such as `DecodingError` and move them into nested properties such as `data` and `error`?

File: packages/signature-controller/src/types.ts [hunk 11]
```diff
@@ -71,11 +77,29 @@ export type MessageParamsTyped = MessageParams & {
...
+      error: {
+        message: string;
+        type: string;
+      };
+    }
+  | 'IN_PROGRESS'
```
Comment (matthewwalsh0) [inline packages/signature-controller/src/types.ts:100]: As mentioned elsewhere, is it cleaner to track this via an explicit `decodingLoading` property on the `SignatureRequest`?

File: packages/signature-controller/src/types.ts (hunk 9 above)
Comment (matthewwalsh0) [inline packages/signature-controller/src/types.ts:86]: JSDoc for all types and properties?

File: packages/signature-controller/src/types.ts [hunk 12]
```diff
@@ -109,6 +133,9 @@ type SignatureRequestBase = {
 
   /** Version of the signTypedData request. */
   version?: SignTypedDataVersion;
+
+  /** Response from message decoding api. */
+  decodedRequest?: DecodedRequestInfo;
```
Comment (matthewwalsh0) [inline packages/signature-controller/src/types.ts:138]: Minor, alphabetical?

Comment (matthewwalsh0) [inline packages/signature-controller/src/types.ts:138]: `decodedData` to de-couple from the API and encapsulate the error also?

File: packages/signature-controller/src/SignatureController.test.ts (hunk 1 above)
Comment (matthewwalsh0) [reply packages/signature-controller/src/SignatureController.test.ts:930]: Is the issue that we're awaiting `newUnsignedTypeMessage` but that won't resolve unless we mock the keyring controller?

File: packages/signature-controller/src/SignatureController.test.ts [hunk 13]
```diff
@@ -890,6 +911,23 @@ describe('SignatureController', () => {
         ).version,
       ).toBe(SignTypedDataVersion.V3);
     });
+
+    // eslint-disable-next-line jest/no-disabled-tests
+    it.skip('invoke decoding api for permits', async () => {
```
Comment (matthewwalsh0) [inline packages/signature-controller/src/SignatureController.test.ts:916]: Should we have additional unit tests to verify the final decoded response is saved, and also that errors are caught and saved?

Perhaps in a new `Decoding` describe?

File: packages/signature-controller/src/SignatureController.ts [hunk 14]
```diff
@@ -164,6 +166,11 @@ export type SignatureControllerOptions = {
    * Callback to record the duration of code.
    */
   trace?: TraceCallback;
+
+  /**
+   * Api used to get decoding data for permits.
+   */
+  decodingApi?: string;
```
Comment (matthewwalsh0) [inline packages/signature-controller/src/SignatureController.ts:173]: Probably a future concern but we may want a callback such as `isDecodingEnabled` that allows the client to disable it per request.

[the remark this answers is not in the document]
Comment (jpuri) [reply packages/signature-controller/src/SignatureController.ts:173]: Fixed.

[the remark this answers is not in the document]
Comment (jpuri) [reply packages/signature-controller/src/SignatureController.ts:173]: fixed

File: packages/signature-controller/src/SignatureController.ts [hunk 15]
```diff
@@ -178,15 +185,23 @@ export class SignatureController extends BaseController<
 
   #trace: TraceCallback;
 
+  #decodingApi?: string;
```
[the remark this answers is not in the document]
Comment (jpuri) [reply packages/signature-controller/src/SignatureController.ts:188]: fixed

File: packages/signature-controller/src/SignatureController.ts [hunk 16]
```diff
@@ -178,15 +185,23 @@ export class SignatureController extends BaseController<
...
    * @param options.state - Initial state to set on this controller.
    * @param options.trace - Callback to generate trace information.
+   * @param options.decodingApi - Api used to get decoded data for permits.
    */
-  constructor({ messenger, state, trace }: SignatureControllerOptions) {
+  constructor({
+    messenger,
+    state,
+    trace,
+    decodingApi,
```
[the remark this answers is not in the document]
Comment (jpuri) [reply packages/signature-controller/src/SignatureController.ts:203]: fixed

File: packages/signature-controller/src/SignatureController.ts (hunk 3 above)
Comment (jpuri) [reply packages/signature-controller/src/SignatureController.ts:915]: Updated.

File: packages/signature-controller/src/SignatureController.ts (hunk 5 above)
Comment (jpuri) [reply packages/signature-controller/src/SignatureController.ts:935]: I preferred then/catch here to help us keep this method async while processing signature request

File: packages/signature-controller/src/types.ts (hunk 7 above)
Comment (jpuri) [reply packages/signature-controller/src/types.ts:20]: Fixed

File: packages/signature-controller/src/types.ts (hunk 8 above)
Comment (jpuri) [reply packages/signature-controller/src/types.ts:81]: I removed this

File: packages/signature-controller/src/types.ts (hunk 9 above)
Comment (jpuri) [reply packages/signature-controller/src/types.ts:86]: Fixed.

File: packages/signature-controller/src/types.ts (hunk 11 above)
Comment (jpuri) [reply packages/signature-controller/src/types.ts:100]: Fixed

File: packages/signature-controller/src/types.ts (hunk 12 above)
Comment (jpuri) [reply packages/signature-controller/src/types.ts:141]: fixed

File: packages/signature-controller/src/SignatureController.test.ts (hunk 1 above)
Comment (jpuri) [reply packages/signature-controller/src/SignatureController.test.ts:930]: There are other similar tests here which are working, I think. reason is async call to decoding api.

File: packages/signature-controller/src/SignatureController.ts [hunk 17]
```diff
@@ -880,4 +897,67 @@ export class SignatureController extends BaseController<
...
+    }
+    let decodedRequest: DecodedRequestInfo;
+    try {
+      const { primaryType } = JSON.parse(request.params?.[1] ?? '');
+      if (primaryType === 'Permit') {
+        this.#updateMetadata(signatureRequestId, (draftMetadata) => {
+          draftMetadata.decodedRequest = 'IN_PROGRESS';
+        });
+        const { method, origin, params } = request;
+        fetch(`${this.#decodingApi}?chainId=${chainId}`, {
```
[the remark this answers is not in the document]
Comment (jpuri) [reply packages/signature-controller/src/SignatureController.ts]: Make sense, I updated the PR.

File: packages/signature-controller/src/SignatureController.ts [hunk 18]
```diff
@@ -880,4 +897,67 @@ export class SignatureController extends BaseController<
...
+              params?.[0],
+              JSON.parse(
+                convertNumbericValuestoQuotedString(params?.[1]) ?? '',
+              ),
+            ],
+          }),
+          headers: { 'Content-Type': 'application/json' },
+        })
+          .then((response) => {
+            return response.json();
```
[the remark this answers is not in the document]
Comment (jpuri) [reply packages/signature-controller/src/SignatureController.ts]: Our error property is consistent with the structure of error returned from the api.

File: packages/signature-controller/src/SignatureController.test.ts (hunk 13 above)
Comment (jpuri) [reply packages/signature-controller/src/SignatureController.test.ts:916]: yep adding those.

File: packages/signature-controller/src/SignatureController.ts (hunk 2 above)
Comment (jpuri) [reply packages/signature-controller/src/SignatureController.ts]: I changed it to check for method `eth_signTypedData_v4`.

File: packages/signature-controller/src/SignatureController.test.ts (hunk 1 above)
Comment (jpuri) [reply packages/signature-controller/src/SignatureController.test.ts:973]: This is now fixed using mocking.

File: packages/signature-controller/src/SignatureController.ts [hunk 19]
```diff
@@ -155,6 +156,11 @@ export type SignatureControllerOptions = {
     // eslint-disable-next-line @typescript-eslint/no-explicit-any
   ) => Promise<any>;
 
+  /**
+   * Api used to get decoding data for permits.
```
Comment (matthewwalsh0) [inline packages/signature-controller/src/SignatureController.ts:160]:
```suggestion
   * URL of API to retrieve decoding data for typed requests.
```

File: packages/signature-controller/src/SignatureController.ts [hunk 20]
```diff
@@ -185,8 +193,14 @@ export class SignatureController extends BaseController<
    * @param options.messenger - The restricted controller messenger for the sign controller.
    * @param options.state - Initial state to set on this controller.
    * @param options.trace - Callback to generate trace information.
+   * @param options.decodingApiUrl - Api used to get decoded data for permits.
```
Comment (matthewwalsh0) [inline packages/signature-controller/src/SignatureController.ts:196]: Minor, but since we now type the option properties, do we need this constructor JSDoc also?

File: packages/signature-controller/src/types.ts [hunk 21]
```diff
@@ -7,12 +7,18 @@ export type OriginalRequest = {
...
+  method?: string;
+
   /** ID of the network client associated with the request. */
   networkClientId?: string;
 
   /** Source of the client request. */
   origin?: string;
 
+  /** Parameters in signature request */
+  params: string[];
```
Comment (matthewwalsh0) [inline packages/signature-controller/src/types.ts:20]: Are we sure these are always strings or should we use `unknown`?

File: packages/signature-controller/src/types.ts [hunk 22]
```diff
@@ -71,15 +77,41 @@ export type MessageParamsTyped = MessageParams & {
         primaryType: string;
         message: Json;
       };
-
   /** Version of the signTypedData request. */
   version?: string;
 };
 
+/** Information about various state changes returned by decoding api. */
+type DecodingDataStateChanges = {
```
Comment (matthewwalsh0) [inline packages/signature-controller/src/types.ts:85]: Rather than defining this as an array type, is it more flexible to define a singular type such as `DecodingDataStateChange` and then set the `stateChanges` property as an array of it?

File: packages/signature-controller/src/types.ts [hunk 23]
```diff
@@ -71,15 +77,41 @@ export type MessageParamsTyped = MessageParams & {
...
+  assetType: string;
+  changeType: string;
+  address: string;
+  amount: string;
+  contractAddress: string;
```
Comment (matthewwalsh0) [inline packages/signature-controller/src/types.ts:90]: Should we use the `Hex` type for some of these?

File: packages/signature-controller/src/types.ts [hunk 24]
```diff
@@ -71,15 +77,41 @@ export type MessageParamsTyped = MessageParams & {
...
+}[];
+
+/** Error details for unfulfilled the decoding request. */
+type DecodingDataError = {
+  message: string;
```
Comment (matthewwalsh0) [inline packages/signature-controller/src/types.ts:95]: One for the API potentially, but is it more flexible to return error codes to let the UI generate suitable messages and translate them?

**Review by matthewwalsh0 — CHANGES_REQUESTED** [review]

File: packages/signature-controller/src/types.ts (hunk 21 above)
Comment (jpuri) [reply packages/signature-controller/src/types.ts:20]: They are string AFAIK

File: packages/signature-controller/src/SignatureController.ts (hunk 20 above)
Comment (jpuri) [reply packages/signature-controller/src/SignatureController.ts:196]: Constructor JS docs are existing, would you suggest removing them all ?

**Comment by jpuri** [comment]
Hey @matthewwalsh0 : I have a confusion here:
> I just mean we can write it more naturally with await if we mark the function as async, it shouldn't change the current flow functionally.

I actually do not want the flow to wait for decoding api to return.

File: packages/signature-controller/src/SignatureController.test.ts [hunk 25]
```diff
@@ -890,6 +913,64 @@ describe('SignatureController', () => {
         ).version,
       ).toBe(SignTypedDataVersion.V3);
     });
+
+    // eslint-disable-next-line jest/no-disabled-tests
+    it('invoke getDecodingData to get decoding data', async () => {
```
[the remark this answers is not in the document]
Comment (jpuri) [reply packages/signature-controller/src/SignatureController.test.ts:918]: Done

File: packages/signature-controller/src/SignatureController.test.ts [hunk 26]
```diff
@@ -890,6 +913,64 @@ describe('SignatureController', () => {
         ).version,
       ).toBe(SignTypedDataVersion.V3);
     });
+
+    // eslint-disable-next-line jest/no-disabled-tests
```
[the remark this answers is not in the document]
Comment (jpuri) [reply packages/signature-controller/src/SignatureController.test.ts:917]: Done

File: packages/signature-controller/src/utils/decoding-api.ts [hunk 27]
```diff
@@ -0,0 +1,59 @@
...
+
+/**
+ * The function calls decoding api for typed signature V4 requests and returns the result.
+ *
+ * @param request - Signature request.
+ * @param chainId - chainId of network of signature request.
+ * @param decodingApiUrl - URL of decoding api.
+ * @returns Promise that resolved to give decoded data.
+ */
+export async function getDecodingData(
```
[the remark this answers is not in the document]
Comment (jpuri) [reply packages/signature-controller/src/utils/decoding-api.ts]: Done

File: packages/signature-controller/src/types.ts (hunk 23 above)
Comment (jpuri) [reply packages/signature-controller/src/types.ts:96]: According to api [doc](https://www.notion.so/metamask-consensys/Signature-API-Documentation-fb31344c77454671ad427c5b88f06b84) it is string types.

File: packages/signature-controller/src/types.ts [hunk 28]
```diff
@@ -71,15 +77,41 @@ export type MessageParamsTyped = MessageParams & {
...
+  type: string;
```
[the remark this answers is not in the document]
Comment (jpuri) [reply packages/signature-controller/src/types.ts:105]: I am sure of all possible values here.

File: packages/signature-controller/src/types.ts [hunk 29]
```diff
@@ -71,15 +77,41 @@ export type MessageParamsTyped = MessageParams & {
...
       };
-
   /** Version of the signTypedData request. */
   version?: string;
 };
 
+/** Information about various state changes returned by decoding api. */
+type DecodingDataStateChanges = {
+  assetType: string;
+  changeType: string;
```
[the remark this answers is not in the document]
Comment (jpuri) [reply packages/signature-controller/src/types.ts]: done 👍

File: packages/signature-controller/src/types.ts (hunk 22 above)
Comment (jpuri) [reply packages/signature-controller/src/types.ts]: done

File: packages/signature-controller/src/SignatureController.ts (hunk 5 above)
Comment (jpuri) [reply packages/signature-controller/src/SignatureController.ts]: I actually do not want the flow to wait for decoding api to return.

File: packages/signature-controller/src/types.ts [hunk 30]
```diff
@@ -71,15 +77,41 @@ export type MessageParamsTyped = MessageParams & {
...
+
+/** Error details for unfulfilled the decoding request. */
+type DecodingDataError = {
+  message: string;
+  type: string;
+};
+
+/** Decoding data about typed sign V4 signature request. */
+export type DecodingData = {
+  stateChanges: DecodingDataStateChanges | null;
```
[the remark this answers is not in the document]
Comment (jpuri) [reply packages/signature-controller/src/types.ts:110]: api returns `null` in case of error

File: packages/signature-controller/src/utils/decoding-api.ts [hunk 31]
```diff
@@ -0,0 +1,59 @@
+import type { OriginalRequest } from '../types';
+import { convertNumericValuesToQuotedString } from './normalize';
+
+export const DECODING_API_ERRORS = {
```
[the remark this answers is not in the document]
Comment (jpuri) [reply packages/signature-controller/src/utils/decoding-api.ts:4]: I prefer object types for strings, any specific reason to prefer enums ?

File: packages/signature-controller/src/utils/decoding-api.ts [hunk 32]
```diff
@@ -0,0 +1,59 @@
...
+/**
+ * The function calls decoding api for typed signature V4 requests and returns the result.
+ *
+ * @param request - Signature request.
+ * @param chainId - chainId of network of signature request.
+ * @param decodingApiUrl - URL of decoding api.
+ * @returns Promise that resolved to give decoded data.
+ */
+export async function getDecodingData(
+  request: OriginalRequest,
```
[the remark this answers is not in the document]
Comment (jpuri) [reply packages/signature-controller/src/utils/decoding-api.ts:18]: Its only 3 args thus I think its ok, also I am not sure if explicit return type is important when it can be inferred

File: packages/signature-controller/src/utils/decoding-api.ts [hunk 33]
```diff
@@ -0,0 +1,59 @@
...
+  chainId: string,
+  decodingApiUrl?: string,
+) {
+  if (!decodingApiUrl) {
+    return undefined;
+  }
+  try {
+    const { method, origin, params } = request;
+    if (request.method === 'eth_signTypedData_v4') {
```
[the remark this answers is not in the document]
Comment (jpuri) [reply packages/signature-controller/src/utils/decoding-api.ts:27]: done

[the remark this answers is not in the document]
Comment (jpuri) [reply packages/signature-controller/src/utils/decoding-api.ts:27]: it is almost same, not really early return

File: packages/signature-controller/src/utils/decoding-api.ts [hunk 34]
```diff
@@ -0,0 +1,59 @@
...
+ *
+ * @param request - Signature request.
+ * @param chainId - chainId of network of signature request.
+ * @param decodingApiUrl - URL of decoding api.
+ * @returns Promise that resolved to give decoded data.
+ */
+export async function getDecodingData(
+  request: OriginalRequest,
+  chainId: string,
+  decodingApiUrl?: string,
```
[the remark this answers is not in the document]
Comment (jpuri) [reply packages/signature-controller/src/utils/decoding-api.ts:20]: We have this check in place
```
if (!decodingApiUrl) {
    return undefined;
  }
```

[the remark this answers is not in the document]
Comment (jpuri) [reply packages/signature-controller/src/utils/decoding-api.ts:20]: It is optional for UI to pass this URL, thus we need this check here. In mobile we still do not have this implementation in place.

File: packages/signature-controller/src/utils/decoding-api.test.ts [hunk 35]
```diff
@@ -0,0 +1,93 @@
+import nock from 'nock';
+import nodeFetch from 'node-fetch';
+
+import type { OriginalRequest } from '../types';
+import { getDecodingData } from './decoding-api';
+
+// eslint-disable-next-line @typescript-eslint/ban-ts-comment
+// @ts-ignore-next-line
+global.fetch = nodeFetch;
```
[the remark this answers is not in the document]
Comment (jpuri) [reply packages/signature-controller/src/utils/decoding-api.test.ts:9]: done

File: packages/signature-controller/src/SignatureController.test.ts [hunk 36]
```diff
@@ -890,6 +913,64 @@ describe('SignatureController', () => {
...
+
+      expect(controller.state.signatureRequests[ID_MOCK].decodingLoading).toBe(
+        false,
+      );
+      expect(
+        controller.state.signatureRequests[ID_MOCK].decodingData,
+      ).toStrictEqual(MOCK_STATE_CHANGES);
+    });
+
+    it('correctly set decoding data if getDecodingData fails', async () => {
```
[the remark this answers is not in the document]
Comment (jpuri) [reply packages/signature-controller/src/SignatureController.test.ts]: done

File: packages/signature-controller/src/SignatureController.ts (hunk 5 above)
Comment (matthewwalsh0) [reply packages/signature-controller/src/SignatureController.ts]: I understand that, but this method can still be `async` without blocking the flow, we just wouldn't `await` when we call it.

File: packages/signature-controller/src/utils/decoding-api.ts (hunk 32 above)
[the remark this answers is not in the document]
Comment (matthewwalsh0) [reply packages/signature-controller/src/utils/decoding-api.ts:20]: It's okay for the time being but would make it more future proof and extensible.

File: packages/signature-controller/src/utils/decoding-api.ts [hunk 37]
```diff
@@ -0,0 +1,61 @@
+import { EthMethod } from '@metamask/keyring-api';
```
Comment (matthewwalsh0) [inline packages/signature-controller/src/utils/decoding-api.ts:1]: If we can't source that constant without this value, then better we define it ourselves to minimise dependencies.

File: packages/signature-controller/src/types.ts [hunk 38]
```diff
@@ -71,15 +77,50 @@ export type MessageParamsTyped = MessageParams & {
...
       };
-
   /** Version of the signTypedData request. */
   version?: string;
 };
 
+/** Information about a single state change returned by decoding api. */
+type DecodingDataStateChange = {
+  assetType: string;
+  changeType:
```
Comment (matthewwalsh0) [inline packages/signature-controller/src/types.ts:87]: Minor, perhaps a separate `DecodingDataChangeType`?

File: packages/signature-controller/src/types.ts [hunk 39]
```diff
@@ -71,15 +77,50 @@ export type MessageParamsTyped = MessageParams & {
...
+/** Error details for unfulfilled the decoding request. */
+type DecodingDataError = {
+  message: string;
+  type: string;
+};
+
+/** Decoding data about typed sign V4 signature request. */
+export type DecodingData = {
+  stateChanges: DecodingDataStateChanges | null;
+  error: DecodingDataError;
```
Comment (matthewwalsh0) [inline packages/signature-controller/src/types.ts:111]: Is this not optional as successful responses will have this as `undefined`?

**Review by matthewwalsh0 — CHANGES_REQUESTED** [review]
O

File: packages/signature-controller/src/utils/decoding-api.ts (hunk 37 above)
Comment (jpuri) [reply packages/signature-controller/src/utils/decoding-api.ts:1]: done

File: packages/signature-controller/src/types.ts (hunk 38 above)
Comment (jpuri) [reply packages/signature-controller/src/types.ts:87]: fixed

File: packages/signature-controller/src/constants.ts [hunk 40]
```diff
@@ -0,0 +1,8 @@
+export const EthMethod = {
```
Comment (matthewwalsh0) [inline packages/signature-controller/src/constants.ts:1]: Minor, this feels more like an enum as it's a fixed set of values to identify a type of something.

Which would also mean it could go in `types.ts`.

But we can do this later since it's not exported.

File: packages/signature-controller/package.json [hunk 41]
```diff
@@ -58,6 +58,7 @@
   "devDependencies": {
     "@metamask/approval-controller": "^7.1.1",
     "@metamask/auto-changelog": "^3.4.4",
+    "@metamask/keyring-api": "^8.1.3",
```
Comment (matthewwalsh0) [inline packages/signature-controller/package.json:61]: Apologies, we forgot to remove this.

**Review by matthewwalsh0 — CHANGES_REQUESTED** [review]
Happy to approve once the `keyring-api` dependency is removed.

File: packages/signature-controller/src/constants.ts (hunk 40 above)
Comment (jpuri) [reply packages/signature-controller/src/constants.ts:4]: I tried that but it is giving build issues somehow.

File: packages/signature-controller/package.json (hunk 41 above)
Comment (jpuri) [reply packages/signature-controller/package.json]: done

**Review by matthewwalsh0 — APPROVED** [review]
